R/devmode.R

```diff
@@ -0,0 +1,251 @@
+#'     # !testthat::is_testing()
+#'     !identical(Sys.getenv("TESTTHAT"), "true")
+#' }
+#' ```
+#'
+#' 2. `get_devmode_option(option, default, on, message)`:
+#' This function should return the global `option` when it is set, similar to `getOption(option, default)`. When the global option `option` is not set, this function should return the default value (`default`) when `in_devmode()` is `FALSE` and the default Dev Mode value (`on`) when `in_devmode()` is `TRUE`. We strongly recommend displaying a message (`message`) to the developer once every 8 hours using `rlang::inform(message, .frequency = "regularly", .frequency_id = message)` if returning the `on` default Dev Mode option value. This will keep the author up to date as to which behaviors are being altered. To allow developers a chance to disable Dev Mode messages, the message may be skipped if `getOption("shiny.devmode.verbose", TRUE)` returns `FALSE`.
+#'
+#' ```r
+#' get_devmode_option <- function(option, default = NULL, on = NULL, message = NULL) {
+#'   if (!in_devmode()) {
+#'     # Dev Mode disabled, act like `getOption()`
+#'     return(getOption(option, default = default))
+#'   }
+#'   # Notify developer
+#'   if (getOption("shiny.devmode.verbose", TRUE) && !is.null(message)) {
+#'     rlang::inform(message = message, .frequency = "regularly", .frequency_id = message)
+#'   }
+#'   # Return Dev Mode default value `on`
+#'   getOption(option, default = on)
+#' }
+#' ```
+#'
+#' The remaining functions in this file are used for author convenience and are not recommended for all situations.
+#' @export
+#' @examples
+#' in_devmode() # TRUE/FALSE?
+#'
+in_devmode <- function() {
+  isTRUE(getOption("shiny.devmode", FALSE)) &&
+    # !testthat::is_testing()
+    !identical(Sys.getenv("TESTTHAT"), "true")
+}
+
+#' @describeIn devmode Temporarily set Shiny Developer Mode and Developer message verbosity
+#' @param code Code to execute with the temporary Dev Mode options set
+#' @export
+#' @examples
+#' # Execute code in a temporary shiny dev mode
+#' with_devmode(TRUE, in_devmode()) # TRUE
+#'
+with_devmode <- function(
+  devmode,
+  code,
+  verbose = getOption("shiny.devmode.verbose", TRUE)
+) {
+  withr::with_options(
+    list(
+      shiny.devmode = devmode,
+      shiny.devmode.verbose = verbose
+    ),
+    code
+  )
+}
+
+
+#' @describeIn devmode If Shiny Developer Mode and verbosity are enabled, displays a message once every 8 hrs (by default)
+#' @inheritParams rlang::inform
+#' @param ... Parameters passed to [rlang::inform()]
+devmode_inform <- function(
+  message,
+  .frequency = "regularly",
+  .frequency_id = message,
+  ...
+) {
+
+  if (!(
+    in_devmode() &&
+      isTRUE(getOption("shiny.devmode.verbose", TRUE))
+  )) {
+    return()
+  }
+  if (is.null(message)) {
+    return()
+  }
+
+  rlang::inform(
+    message = paste0("shiny devmode - ", message),
+    .frequency = .frequency,
+    .frequency_id = .frequency_id,
+    ...
+  )
+}
```

Comment:
The formatting needs to be a bit different: no `\n` between the message and the "once every 8 hours" part, and there should be a `\n` after that.

![image](https://user-images.githubusercontent.com/86978/102110145-6f199d00-3dfa-11eb-912e-f5db447e31b1.png)

Reply:
https://github.com/r-lib/rlang/pull/1076

Reply:
Closed in https://github.com/r-lib/rlang/commit/6b78803e7dff1b75d0b9d9587f20b0d5f37368fa